docs: Draft documentation and user guide for mcp otel telemetry (#338)
apollographql/apollo-mcp-server: docs/source/telemetry.mdx

---
title: OpenTelemetry Integration
---

AI agents create unpredictable usage patterns and complex request flows that are hard to monitor with traditional methods. The Apollo MCP Server's OpenTelemetry integration provides the visibility you need to run a reliable service for AI agents.

> ## **What You Can Monitor**

Reviewer: ```suggestion
## What you can monitor
```

- **Agent behavior**: Which tools and operations are used most frequently
- **Performance**: Response times and bottlenecks across tool executions and GraphQL operations  
- **Reliability**: Error rates, failed operations, and request success patterns
- **Distributed request flows**: Complete traces from agent request through your Apollo Router and subgraphs, with automatic trace context propagation

> ## **How It Works**

Reviewer: ```suggestion
## How it works
```

The server exports metrics, traces, and events using the OpenTelemetry Protocol (OTLP), ensuring compatibility with your existing observability stack and seamless integration with other instrumented Apollo services.

> ## Usage Guide

Reviewer: ```suggestion
## Usage guide
```

### **Quick Start: Local Development**

The fastest way to see Apollo MCP Server telemetry in action is with a local setup that requires only Docker.

**🚀 5-minute setup:**
1. **Start local observability stack**: `docker run -p 3000:3000 -p 4317:4317 -p 4318:4318 --rm -ti grafana/otel-lgtm`
2. **Add telemetry config** to your `config.yaml`:
   ```yaml
   telemetry:
     exporters:
       metrics:
         otlp:
           endpoint: "http://localhost:4318/v1/metrics"
           protocol: "http/protobuf"
       tracing:
         otlp:
           endpoint: "http://localhost:4318/v1/traces" 
           protocol: "http/protobuf"
   ```
3. **Restart your MCP server** with the updated config

> 4. **Open Grafana** at http://localhost:3000 (admin/admin) and explore your telemetry data

Reviewer: ```suggestion
1. Open Grafana at `http://localhost:3000` and explore your telemetry data. Default credentials are username `admin` with password `admin`.
```

For detailed steps and dashboard examples, see the [complete Grafana setup guide](guides/telemetry-grafana.mdx).

> ### **Production Deployment**

Reviewer: ```suggestion
### Production deployment
```

For production environments, configure your MCP server to send telemetry to any OTLP-compatible backend. The Apollo MCP Server uses standard OpenTelemetry protocols, ensuring compatibility with all major observability platforms.

> #### **Configuration Example**

Reviewer: ```suggestion
#### Configuration example
```

```yaml
telemetry:
  service_name: "mcp-server-prod"      # Custom service name
  exporters:
    metrics:
      otlp:
        endpoint: "https://your-metrics-endpoint"
        protocol: "http/protobuf"       # or "grpc"
    tracing:
      otlp:
        endpoint: "https://your-traces-endpoint" 
        protocol: "http/protobuf"
```

> #### **Observability Platform Integration**

Reviewer: ```suggestion
#### Observability platform integration
```

The MCP server works with any OTLP-compatible backend. Consult your provider's documentation for specific endpoint URLs and authentication:

> - **[Datadog OTLP Integration](https://docs.datadoghq.com/tracing/setup_overview/open_standards/otlp_ingest_in_datadog/)** - Native OTLP support
> - **[New Relic OpenTelemetry](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/)** - Direct OTLP ingestion
> - **[AWS Observability](https://aws-otel.github.io/docs/introduction)** - Via AWS Distro for OpenTelemetry
> - **[Grafana Cloud](https://grafana.com/docs/grafana-cloud/send-data/otlp/)** - Hosted Grafana with OTLP
> - **[Honeycomb](https://docs.honeycomb.io/getting-data-in/opentelemetry/)** - OpenTelemetry-native platform
> - **[Jaeger](https://www.jaegertracing.io/docs/1.50/deployment/)** - Self-hosted tracing
> - **[OpenTelemetry Collector](https://opentelemetry.io/docs/collector/deployment/)** - Self-hosted with flexible routing

Reviewer: ```suggestion
- [Datadog OTLP Integration](https://docs.datadoghq.com/tracing/setup_overview/open_standards/otlp_ingest_in_datadog/) - Native OTLP support
- [New Relic OpenTelemetry](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/) - Direct OTLP ingestion
- [AWS Observability](https://aws-otel.github.io/docs/introduction) - Via AWS Distro for OpenTelemetry
- [Grafana Cloud](https://grafana.com/docs/grafana-cloud/send-data/otlp/) - Hosted Grafana with OTLP
- [Honeycomb](https://docs.honeycomb.io/getting-data-in/opentelemetry/) - OpenTelemetry-native platform
- [Jaeger](https://www.jaegertracing.io/docs/1.50/deployment/) - Self-hosted tracing
- [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/deployment/) - Self-hosted with flexible routing
```

> #### **Production Configuration Best Practices**

Reviewer: ```suggestion
#### Production configuration best practices
```

> **Environment and Security:**

Reviewer: ```suggestion
##### Environment and security
```

```yaml
# Set via environment variable
export ENVIRONMENT=production

telemetry:
  service_name: "apollo-mcp-server"
  version: "1.0.0"                     # Version for correlation
  exporters:
    metrics:
      otlp:
        endpoint: "https://secure-endpoint"  # Always use HTTPS
        protocol: "http/protobuf"           # Generally more reliable than gRPC

> ```

Reviewer: I think it would make sense to either introduce or explain this. I don't know which is better, but I feel like we need to explain what this file does for the environment and security settings.

> **Performance Considerations:**

Reviewer: ```suggestion
##### Performance considerations
```

- **Protocol choice**: `http/protobuf` is often more reliable through firewalls and load balancers than `grpc`
- **Batch export**: OpenTelemetry automatically batches telemetry data for efficiency
- **Network timeouts**: Default timeouts are usually appropriate, but monitor for network issues

> **Resource Correlation:**

Reviewer: ```suggestion
##### Resource correlation
```

- The `ENVIRONMENT` variable automatically tags all telemetry with `deployment.environment.name`
- Use consistent `service_name` across all your Apollo infrastructure (Router, subgraphs, MCP server)
- Set `version` to track releases and correlate issues with deployments

> #### **Troubleshooting**

Reviewer: ```suggestion
#### Troubleshooting
```

> **Common Issues:**

Reviewer: ```suggestion
##### Common ssues
```

- **Connection refused**: Verify endpoint URL and network connectivity
- **Authentication errors**: Check if your provider requires API keys or special headers
- **Missing data**: Confirm your observability platform supports OTLP and is configured to receive data
- **High memory usage**: Monitor telemetry export frequency and consider sampling for high-volume environments

> **Verification:**

Reviewer: ```suggestion
##### Verification
```

| `apollo.mcp.list_tools.count` | Counter | Incremented for each `list_tools` request. | (none) |
| `apollo.mcp.get_info.count` | Counter | Incremented for each `get_info` request. | (none) |
| `apollo.mcp.tool.count` | Counter | Incremented for each tool call. | `tool_name`, `success` (bool) |
| `apollo.mcp.tool.duration` | Histogram | Measures the execution duration of each tool call. | `tool_name`, `success` (bool) |
| `apollo.mcp.operation.count`| Counter | Incremented for each downstream GraphQL operation executed by a tool. | `operation.id`, `operation.type` ("persisted_query" or "operation"), `success` (bool) |
| `apollo.mcp.operation.duration`| Histogram | Measures the round-trip duration of each downstream GraphQL operation. | `operation.id`, `operation.type`, `success` (bool) |

In addition to these custom metrics, the server also emits standard [HTTP server metrics](https://opentelemetry.io/docs/specs/semconv/http/http-metrics/) (e.g., `http.server.duration`, `http.server.active_requests`) courtesy of the `axum-otel-metrics` library.

## Emitted Traces

Spans are generated for the following actions:

> *   **Incoming HTTP Requests**: A root span is created for every HTTP request to the MCP server.
> *   **MCP Handler Methods**: Nested spans are created for each of the main MCP protocol methods (`initialize`, `call_tool`, `list_tools`).
> *   **Tool Execution**: `call_tool` spans contain nested spans for the specific tool being executed (e.g., `introspect`, `search`, or a custom GraphQL operation).
> *   **Downstream GraphQL Calls**: The `execute` tool and custom operation tools create child spans for their outgoing `reqwest` HTTP calls, capturing the duration of the downstream request. The `traceparent` and `tracestate` headers are propagated automatically, enabling distributed traces.

Reviewer: ```suggestion
-   **Incoming HTTP Requests**: A root span is created for every HTTP request to the MCP server.
-   **MCP Handler Methods**: Nested spans are created for each of the main MCP protocol methods (`initialize`, `call_tool`, `list_tools`).
-   **Tool Execution**: `call_tool` spans contain nested spans for the specific tool being executed (e.g., `introspect`, `search`, or a custom GraphQL operation).
-   **Downstream GraphQL Calls**: The `execute` tool and custom operation tools create child spans for their outgoing `reqwest` HTTP calls, capturing the duration of the downstream request. The `traceparent` and `tracestate` headers are propagated automatically, enabling distributed traces.
```

This integration provides visibility that is crucial for diagnosing errors, identifying performance bottlenecks, and understanding how different tools are being used.

## Usage Guide

To start exporting telemetry from the Apollo MCP Server, you need an OpenTelemetry Collector to which the MCP server can export metrics and traces to.
These are often run as side cars or agents alongside deployed services and need to be accessible over the network to your instance of the Apollo MCP Server.

Update your config.yaml to configure the MCP Server with the telemetry section. This will point the MCP Server at your collector on startup.

```yaml
telemetry:
  exporters:
    metrics:
      otlp:
        endpoint: "http://localhost:4317"
        protocol: "grpc"
    tracing:
      otlp:
        endpoint: "http://localhost:4317"
        protocol: "grpc"
```
With this configuration, the server will send traces and metrics to an OpenTelemetry Protocol compatible collector running at http://localhost:4317.

> ## Configuration Reference

Reviewer: Agreed. We should do that for sure.

The OpenTelemetry integration is configured via the `telemetry` section in your config.yaml file.

| Parameter | Type | Description | Default |
| :--- | :--- | :--- | :--- |
| service_name | String | The name of the service, which will be included in the telemetry data. | apollo-mcp-server |
| version | String | The version of the service. | The crate version. |

> | exporters | Object | Configuration for the telemetry exporters. | null |

Reviewer 2: Hmm... I'm not sure about the solution. How about we consult with the docs expert, @mabuyo? 😁

`exporters`

| Parameter | Type | Description |
| :--- | :--- | :--- |
| metrics | Object | Configuration for the metrics exporters. |
| tracing | Object | Configuration for the tracing exporters. |

`metrics.otlp` / `tracing.otlp`

| Parameter | Type | Description | Default |
| :--- | :--- | :--- | :--- |
| endpoint | String | The OTLP endpoint to send data to. | http://localhost:4317 |
| protocol | String | The OTLP protocol to use. Supported values are grpc and http/protobuf. | grpc |

## Emitted Metrics

> The server emits the following custom metrics, which are invaluable for monitoring and alerting. All duration metrics are in milliseconds.

Author: Agree, I'll remove the work custom here.
